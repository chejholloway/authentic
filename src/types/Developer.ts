export default interface Developer {
  avatar_url: string;
  login: string;
}
