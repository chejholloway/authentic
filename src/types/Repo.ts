import Owner from './Owner';

/**
 * Represents a GitHub repository's information.
 *
 * This interface defines the structure of information associated with a GitHub repository.
 * It includes details such as the repository's name, owner, whether it's private, and more.
 *
 * @interface Repo
 * @example
 * ```ts
 * const repository: Repo = {
 *   id: 12345,
 *   name: "example-repo",
 *   full_name: "username/example-repo",
 *   private: false,
 *   owner: {
 *     login: "exampleOwner",
 *     id: 123,
 *     // ...owner properties
 *   },
 *   // ...other properties
 * };
 * ```
 */
export default interface Repo {
  id: number;
  node_id: string;
  name: string;
  full_name: string;
  private: boolean;
  owner?: Owner;
}
