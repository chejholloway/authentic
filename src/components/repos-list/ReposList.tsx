import React from "react";
import { Link } from "react-router-dom";
import {
  List,
  ListItem,
  ListItemPrefix,
  Avatar,
  Card,
  Typography,
} from "@material-tailwind/react";
import { Title } from "../../styles/TwinStyles.styles";

import Repo from "../../types/Repo";
import Developer from "../../types/Developer";

interface ReposListProps {
  developer: Developer | undefined;
  repos: Repo[] | undefined;
}

const ReposList: React.FC<ReposListProps> = ({ developer, repos }) => {
  return (
    <div className="text-gray-700">
      <h3 className="text-center">
        <Title className="font-normal">Repositories</Title>
      </h3>
      {repos ? (
        repos.map((repo) => (
          <div key={repo.id}>
            <Link to={repo?.html_url || ""}>
              <Card className="m-2">
                <List>
                  <ListItem>
                    <ListItemPrefix>
                      <Avatar
                        variant="circular"
                        alt={developer?.login}
                        className="outline"
                        src="/assets/images/git-svgrepo-com.svg"
                      />
                    </ListItemPrefix>
                    <Typography
                      variant="small"
                      color="gray"
                      className="font-normal"
                    >
                      {repo.name}
                    </Typography>
                  </ListItem>
                </List>
              </Card>
            </Link>
          </div>
        ))
      ) : (
        <p>No repositories available.</p>
      )}
    </div>
  );
};

export default ReposList;
