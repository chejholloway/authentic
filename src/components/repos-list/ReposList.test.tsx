import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import {
  List,
  ListItem,
  ListItemPrefix,
  Avatar,
  Card,
  Typography,
} from '@material-tailwind/react';
import ReposList from './ReposList';
import Repo from '../../types/Repo';
import Developer from '../../types/Developer';

jest.mock('@material-tailwind/react', () => ({
  Avatar: ({ alt }: { alt: string }) => <img alt={alt} />,
  Typography: ({ children }: { children: React.ReactNode }) => <p>{children}</p>,
  List: ({ children }: { children: React.ReactNode }) => <ul>{children}</ul>,
  ListItem: ({ children }: { children: React.ReactNode }) => <li>{children}</li>,
  ListItemPrefix: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
  Card: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
  Link: ({ children }: { children: React.ReactNode }) => <a>{children}</a>,
}));

describe('<ReposList>', () => {
  test('renders repositories list with data', () => {
    const mockRepos: Repo[] = [
      {
        id: 1,
        name: 'repo1',
      },
      {
        id: 2,
        name: 'repo2',
      },
    ];

    const mockDeveloper: Developer = {
      id: 1,
      login: 'developer1',
      avatar_url: 'developer1-avatar-url',
    };

    const { asFragment } = render(
      <Router>
        <ReposList developer={mockDeveloper} repos={mockRepos} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('renders "No repositories available." when repos are undefined', () => {
    const mockDeveloper: Developer = {
      id: 1,
      login: 'developer1',
      avatar_url: 'developer1-avatar-url',
    };

    const { asFragment } = render(
      <Router>
        <ReposList developer={mockDeveloper} repos={undefined} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
