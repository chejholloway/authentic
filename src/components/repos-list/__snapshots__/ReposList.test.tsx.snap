// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReposList> renders "No repositories available." when repos are undefined 1`] = `
<DocumentFragment>
  <div
    class="text-gray-700"
  >
    <h3
      class="text-center"
    >
      <p>
        Repositories
      </p>
    </h3>
    <p>
      No repositories available.
    </p>
  </div>
</DocumentFragment>
`;

exports[`<ReposList> renders repositories list with data 1`] = `
<DocumentFragment>
  <div
    class="text-gray-700"
  >
    <h3
      class="text-center"
    >
      <p>
        Repositories
      </p>
    </h3>
    <div>
      <a
        href="/"
      >
        <div>
          <ul>
            <li>
              <div>
                <img
                  alt="developer1"
                />
              </div>
              <p>
                repo1
              </p>
            </li>
          </ul>
        </div>
      </a>
    </div>
    <div>
      <a
        href="/"
      >
        <div>
          <ul>
            <li>
              <div>
                <img
                  alt="developer1"
                />
              </div>
              <p>
                repo2
              </p>
            </li>
          </ul>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
